WIP: experimenting with Clang -fbounds-safety

seanm opened the pull request:
I've been playing with [-fbounds-safety](https://clang.llvm.org/docs/BoundsSafety.html), and libusb is a nice small codebase to try it with.

Here are some WIP patches.  A couple of them could be merged soon, the others are just to show and share. There's a lot left to do.

I've already found one place that will require an API change ideally, which is why I created this PR now, so that can be discussed...

seanm committed:
WIP: Added buffer sizing annotations for -fbounds-safety

seanm committed:
Always set pointer and its size together for -fbounds-safety

-fbounds-safety requires that pointer and its size are always set back-to-back with no intervening statements. Before this patch, the `if` statement was in the way.

This changes also makes sure that length is set to 0 if the pointer is NULL.

seanm committed:
WIP: Changed libusb_fill_control_transfer to track length of buffer

The function signature provided a buffer but no associated length.  This is not so safe, and won't compile with -fbounds-safety.

Other similar functions like libusb_fill_bulk_transfer, libusb_fill_bulk_stream_transfer, libusb_fill_interrupt_transfer, etc. all sensibly group both together.

seanm committed:
Always set pointer and its size together for -fbounds-safety

-fbounds-safety requires that pointer and its size are always set back-to-back with no intervening statements.

Before this patch, the loop was incrementing the count without also syncing the pointer, which does not compile.

Added temporary variable to count then set pointer and size back-to-back.

Also removed unneed NULL test since free(NULL) is well-defined to do nothing.

seanm commented:
@tormodvolden @sonatique care to take a look?

seanm commented on libusb/os/events_posix.c:
```diff
@@ -224,16 +224,15 @@ int usbi_alloc_event_data(struct libusb_context *ctx)
+		cnt++;
 
-	fds = calloc(ctx->event_data_cnt, sizeof(*fds));
+	fds = calloc(cnt, sizeof(*fds));
```
Thanks for reviewing.

>Why is -fbounds-safety complaining?

First, I'll just say: there was no bug here, the code was fine with normal C.

-fbounds-safety is more like a different dialect of C, so it has different rules. It was complaining because one of those rules is that when a pointer tagged with `sized_by` or `counted_by` is changed, the count variable tied to it has to change too, and vice versa, and the changes have to be back-to-back. The document I linked explains in detail, but it is of course pretty long.

>how the changes improve something for real, not just to please it?

"just pleasing it" allows the code to be built with -fbound-safety, and then you get memory safety: buffer overruns will always crash, instead of silently overwriting memory. That improves something for real.

tormodvolden commented on libusb/libusb.h:
```diff
@@ -1927,10 +1927,18 @@ static inline void libusb_fill_control_transfer(
 	if (setup)
-		transfer->length = (int) (LIBUSB_CONTROL_SETUP_SIZE
+	{
+		int length = (int) (LIBUSB_CONTROL_SETUP_SIZE
```
Does it also require the helper variable?

seanm replied:
IIRC, yes in this case. Because the compiler needs to see both the pointer field (buffer) and the size field (length) updated back to back with no intervening stuff.

https://clang.llvm.org/docs/BoundsSafety.html#maintaining-correctness-of-bounds-annotations

"Additionally, an explicit bounds annotation such as `int *__counted_by(count) buf` defines a relationship between two variables, buf and count: namely, that buf has count number of elements available. This relationship must hold even after any of these related variables are updated. To this end, the model requires that assignments to buf and count must be side by side, with no side effects between them. This prevents buf and count from temporarily falling out of sync due to updates happening at a distance."

seanm committed:
WIP: bounds safety

seanm committed:
Always set pointer and its size together for -fbounds-safety

Eliminated `remaining` variable in favour of just mutating the `size` parameter, which has to be kept in sync with the `buffer` parameter.

Should be no change in behaviour here.

seanm committed:
Always set pointer and its size together for -fbounds-safety

Use a local variable to count so that setting the pointer and count in the structure can be delayed and done together.

Small behaviour change: move zero initialization to the beginning of the function, as such if early return occurs because of error, these fields will now be zeroed. Shouldn't change much in practice.

seanm committed:
Use const in main() signature

This helps with -fbound-safety because it assumes that const char* are null terminated stings, which is case for main().
